array dimension set by a constant expression

The following program `sample.go` triggers a panic:

```go
package main

const (
    maxNonStarters = 30
    maxBufferSize  = maxNonStarters + 2
)

type reorderBuffer struct {
    rune [maxBufferSize]Properties
}

type Properties struct {
    pos  uint8
    size uint8
}

func main() {
    println("ok")
}   
```

Expected result:
```console
$ go run ./sample.go
ok
```

Got:
```console
$ yaegi ./sample.go
panic: reflect: call of reflect.Value.Interface on zero Value

goroutine 1 [running]:
reflect.valueInterface(0x0, 0x0, 0x0, 0xc00026d101, 0x0, 0xc00018a101)
        /usr/lib/go/src/reflect/value.go:1008 +0x1a0
reflect.Value.Interface(...)
        /usr/lib/go/src/reflect/value.go:1003
github.com/containous/yaegi/interp.nodeType(0xc000264580, 0xc0001724c0, 0xc00027dd00, 0x0, 0x0, 0xc00014c400)
        /home/marc/go/src/github.com/containous/yaegi/interp/type.go:168 +0x6e5
github.com/containous/yaegi/interp.nodeType(0xc000264580, 0xc0001724c0, 0xc00027d900, 0xa, 0xc000282168, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/type.go:539 +0x47f8
github.com/containous/yaegi/interp.(*Interpreter).gta.func1(0xc00027d700, 0x61bb00)
        /home/marc/go/src/github.com/containous/yaegi/interp/gta.go:165 +0x1df5
github.com/containous/yaegi/interp.(*node).Walk(0xc00027d700, 0xc0001c1d20, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:164 +0xc9
github.com/containous/yaegi/interp.(*node).Walk(0xc00027d600, 0xc0001c1d20, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:168 +0x70
github.com/containous/yaegi/interp.(*node).Walk(0xc00027c000, 0xc0001c1d20, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:168 +0x70
github.com/containous/yaegi/interp.(*Interpreter).gta(0xc000264580, 0xc00027c000, 0xc00018a068, 0x4, 0x18, 0xc00018a068, 0x4, 0xc00027c000, 0x0)
        /home/marc/go/src/github.com/containous/yaegi/interp/gta.go:17 +0x127
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc000264580, 0xc0001500f0, 0xe5, 0x2e5, 0xc0001500f0, 0xe5, 0x0, 0x1)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:311 +0x105
main.main()
        /home/marc/go/src/github.com/containous/yaegi/cmd/yaegi/yaegi.go:82 +0x42e
```

--- a/_test/const6.go
+++ b/_test/const6.go
@@ -1,0 +1,22 @@
+package main
+
+const (
+	maxNonStarters = 30
+	maxBufferSize  = maxNonStarters + 2
+)
+
+type reorderBuffer struct {
+	rune [maxBufferSize]Properties
+}
+
+type Properties struct {
+	pos  uint8
+	size uint8
+}
+
+func main() {
+	println(len(reorderBuffer{}.rune))
+}
+
+// Output:
+// 32

--- a/_test/const7.go
+++ b/_test/const7.go
@@ -1,0 +1,19 @@
+package main
+
+import "fmt"
+
+const (
+	a = iota
+	b
+	c
+	d
+)
+
+type T [c]int
+
+func main() {
+	fmt.Println(T{})
+}
+
+// Output:
+// [0 0]

--- a/interp/ast.go
+++ b/interp/ast.go
@@ -75,8 +75,6 @@ const (
 	parenExpr
 	rangeStmt
 	returnStmt
-	rvalueExpr
-	rtypeExpr
 	selectStmt
 	selectorExpr
 	selectorImport
@@ -154,8 +152,6 @@ var kinds = [...]string{
 	parenExpr:        "parenExpr",
 	rangeStmt:        "rangeStmt",
 	returnStmt:       "returnStmt",
-	rvalueExpr:       "rvalueExpr",
-	rtypeExpr:        "rtypeExpr",
 	selectStmt:       "selectStmt",
 	selectorExpr:     "selectorExpr",
 	selectorImport:   "selectorImport",

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -429,7 +429,8 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 					if sc.global {
 						// Do not overload existing symbols (defined in GTA) in global scope
 						sym, _, _ = sc.lookup(dest.ident)
-					} else {
+					}
+					if sym == nil {
 						sym = &symbol{index: sc.add(dest.typ), kind: varSym, typ: dest.typ}
 						sc.sym[dest.ident] = sym
 					}
@@ -509,9 +510,10 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 				if isMapEntry(dest) {
 					dest.gen = nop // skip getIndexMap
 				}
-			}
-			if n.anc.kind == constDecl {
-				iotaValue++
+				if n.anc.kind == constDecl {
+					sc.sym[dest.ident].kind = constSym
+					iotaValue++
+				}
 			}
 
 		case incDecStmt:
@@ -924,7 +926,8 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 		case identExpr:
 			if isKey(n) || isNewDefine(n, sc) {
 				break
-			} else if sym, level, ok := sc.lookup(n.ident); ok {
+			}
+			if sym, level, ok := sc.lookup(n.ident); ok {
 				// Found symbol, populate node info
 				n.typ, n.findex, n.level = sym.typ, sym.index, level
 				if n.findex < 0 {
@@ -941,11 +944,6 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 					case n.ident == "nil":
 						n.kind = basicLit
 					case sym.kind == binSym:
-						if sym.rval.IsValid() {
-							n.kind = rvalueExpr
-						} else {
-							n.kind = rtypeExpr
-						}
 						n.typ = sym.typ
 						n.rval = sym.rval
 					case sym.kind == bltnSym:
@@ -1161,10 +1159,8 @@ func (interp *Interpreter) cfg(root *node, pkgID string) ([]*node, error) {
 				pkg := n.child[0].sym.typ.path
 				if s, ok := interp.binPkg[pkg][name]; ok {
 					if isBinType(s) {
-						n.kind = rtypeExpr
 						n.typ = &itype{cat: valueT, rtype: s.Type().Elem()}
 					} else {
-						n.kind = rvalueExpr
 						n.typ = &itype{cat: valueT, rtype: s.Type(), untyped: isValueUntyped(s)}
 						n.rval = s
 					}
@@ -1587,7 +1583,7 @@ func isBinType(v reflect.Value) bool { return v.IsValid() && v.Kind() == reflect
 // isType returns true if node refers to a type definition, false otherwise
 func (n *node) isType(sc *scope) bool {
 	switch n.kind {
-	case arrayType, chanType, funcType, interfaceType, mapType, structType, rtypeExpr:
+	case arrayType, chanType, funcType, interfaceType, mapType, structType:
 		return true
 	case parenExpr, starExpr:
 		if len(n.child) == 1 {

--- a/interp/gta.go
+++ b/interp/gta.go
@@ -21,6 +21,7 @@ func (interp *Interpreter) gta(root *node, rpath, pkgID string) ([]*node, error)
 		switch n.kind {
 		case constDecl:
 			iotaValue = 0
+			_, err = interp.cfg(n, pkgID)
 
 		case blockStmt:
 			if n != root {
@@ -63,7 +64,9 @@ func (interp *Interpreter) gta(root *node, rpath, pkgID string) ([]*node, error)
 				if typ.isBinMethod {
 					typ = &itype{cat: valueT, rtype: typ.methodCallType(), isBinMethod: true}
 				}
-				sc.sym[dest.ident] = &symbol{kind: varSym, global: true, index: sc.add(typ), typ: typ, rval: val}
+				if sc.sym[dest.ident] == nil {
+					sc.sym[dest.ident] = &symbol{kind: varSym, global: true, index: sc.add(typ), typ: typ, rval: val}
+				}
 				if n.anc.kind == constDecl {
 					sc.sym[dest.ident].kind = constSym
 					iotaValue++

--- a/interp/gta.go
+++ b/interp/gta.go
@@ -21,6 +21,8 @@ func (interp *Interpreter) gta(root *node, rpath, pkgID string) ([]*node, error)
 		switch n.kind {
 		case constDecl:
 			iotaValue = 0
+			// Early parse of constDecl subtree, to compute all constant
+			// values which may be necessary in further declarations.
 			_, err = interp.cfg(n, pkgID)
 
 		case blockStmt:

--- a/interp/interp_eval_test.go
+++ b/interp/interp_eval_test.go
@@ -78,11 +78,11 @@ func TestEvalBuiltin(t *testing.T) {
 	i := interp.New(interp.Options{})
 	runTests(t, i, []testCase{
 		{src: `a := []int{}; a = append(a, 1); a`, res: "[1]"},
-		{src: `a := []int{1}; a = append(a, 2, 3); a`, res: "[1 2 3]"},
-		{src: `a := []int{1}; b := []int{2, 3}; a = append(a, b...); a`, res: "[1 2 3]"},
+		{src: `b := []int{1}; b = append(a, 2, 3); b`, res: "[1 2 3]"},
+		{src: `c := []int{1}; d := []int{2, 3}; c = append(c, d...); c`, res: "[1 2 3]"},
 		{src: `string(append([]byte("hello "), "world"...))`, res: "hello world"},
-		{src: `a := "world"; string(append([]byte("hello "), a...))`, res: "hello world"},
-		{src: `a := []byte("Hello"); copy(a, "world"); string(a)`, res: "world"},
+		{src: `e := "world"; string(append([]byte("hello "), e...))`, res: "hello world"},
+		{src: `f := []byte("Hello"); copy(f, "world"); string(f)`, res: "world"},
 	})
 }
 

--- a/interp/value.go
+++ b/interp/value.go
@@ -90,9 +90,6 @@ func genValue(n *node) func(*frame) reflect.Value {
 			v = reflect.ValueOf(n.val)
 		}
 		return func(f *frame) reflect.Value { return v }
-	case rvalueExpr:
-		v := n.rval
-		return func(f *frame) reflect.Value { return v }
 	default:
 		if n.rval.IsValid() {
 			v := n.rval

